File: msgpack/repeat/sub.go
package main

import (
	_ "encoding/json"
	"fmt"
	zmq "github.com/taka-wang/zmq3"
	"github.com/ugorji/go/codec"
	_ "time"
)

var (
	mh = &codec.MsgpackHandle{RawToString: true}
)

type CmdHeader struct {
	Receiver string `json:"receiver"`
	Sender   string `json:"sender"`
	Version  string `json:"version"`
	Tid      int    `json:"tid"`
	Method   string `json:"method"`
}

type MbTcpHeader struct {
	IP   string `json:"ip"`
	Port int    `json:"port"`
	ID   int    `json:"id"`
}

type MbWriteRequest struct {
	Code     int    `json:"code"`
	Register int    `json:"register"`
	Value    string `json:"value"`
	Type     string `json:"type"`
	Alias    string `json:"alias"`
}

type MbTcpSingleWriteReq struct {
	CmdHeader
	MbTcpHeader
	MbWriteRequest
}

type MbTcpMultipleWriteReq struct {
	CmdHeader
	MbTcpHeader
	Requests []MbWriteRequest `json:"requests"`
}

func main() {
	sub()
}

func sub() {
	receiver, _ := zmq.NewSocket(zmq.SUB)
	defer receiver.Close()
	receiver.Bind("ipc:///tmp/dummy")

	filter := "mbtcp"
	receiver.SetSubscribe(filter) // filter frame 1

	for {
		msg, _ := receiver.RecvMessage(0) // frame len: 2

		// unpack
		var s MbTcpMultipleWriteReq
		dec := codec.NewDecoderBytes([]byte(msg[1]), mh)
		err := dec.Decode(&s)

		if err != nil {
			fmt.Println(err) // unmarshal from json string failed
		} else {
			fmt.Println(msg[0])
			fmt.Println(s)
		}
	}
}

File: proto/composition/sub.c
// compile: gcc type2.pb-c.h type2.pb-c.c sub.c -lzmq -lczmq -lprotobuf-c  -o subc

#include <czmq.h>
#include "type2.pb-c.h"

int main (int argc, char *argv [])
{
    zctx_t *context = zctx_new ();
    void *subscriber = zsocket_new (context, ZMQ_SUB);
    zsocket_bind (subscriber, "ipc:///tmp/dummy");

    zsocket_set_subscribe (subscriber, "mbtcp"); // set filter
    
    while (!zctx_interrupted) {
        zmsg_t *msg = zmsg_recv(subscriber);
        zframe_t *first = zmsg_pop(msg);
        zframe_t *payload = zmsg_pop(msg);
        char *first_buffer = zframe_strdup(first);
        char *payload_buffer = zframe_strdup(payload);
        printf("method: %s\n", first_buffer);
        //printf("payload: %s\n", payload_buffer);

        // unpack
        MbTcpSingleWriteReq *command;
        CmdHeader *cmd_header;
        MbTcpHeader *mb_tcp_header;
        MbWriteRequest *mb_write_request;

        unsigned len = zframe_size(payload);

        command = mb_tcp_single_write_req__unpack(NULL, len, payload_buffer);
        cmd_header       = command->cmd_header;
        mb_tcp_header    = command->mb_tcp_header;
        mb_write_request = command->mb_write_request;
        
        if (command == NULL) {
            fprintf(stderr, "error unpacking incoming message\n");
            break;
        }

        printf("Recv: %s %s %d %d PRId64 %s\n", cmd_header->receiver, cmd_header->sender, cmd_header->version, cmd_header->tid, cmd_header->method);
        printf("Recv: ip: %s, port: %d, id: %d\n", mb_tcp_header->ip, mb_tcp_header->port, mb_tcp_header->id);
        printf("Recv: %d %d %s %s %s\n", mb_write_request->code, mb_write_request->register_, mb_write_request->value, mb_write_request->type, mb_write_request->alias);

        //zmsg_dump(msg);
        
        //cleanup
        mb_tcp_single_write_req__free_unpacked(command, NULL); // free unpacked command
        zframe_destroy(&first);
        zframe_destroy(&payload);
        zmsg_destroy(&msg );
    }
    
    // cleanup
    zctx_destroy(&context);
    return 0;
}